In [1]:
from IPython.display import display, Markdown, Latex
def printmd(s):
    display(Markdown(s))

# Exercise 4.1


I) $V_2^P(HH) = max(-12, \frac{2}{5}(0 + 0)) = 0$

$V_2^P(HT) = V_2^P(TH) = max(4 - 4, \frac{2}{5} ((4 - 8)^+ + (4 - 2)^+) ) = 0.8$

$V_2^P(TT) = 3, V_1^P(H) = 0.32, V_1^P(T) = 2$

$V_0^P = 9.28$

II) $V_0^C = 5$

III)

IV) 

a > b and d > c OR d > a and b > c

<=>

$\max(a, b) + max(c, d) \ge \max(a + c, b + d)$

$$V_n^S = \max\Big(g_S(S_n),  \frac{\overline{p} V_{n + 1}^S + \overline{q} V_{n + 1}^S}{1 + r}\Big)$$
By induction (by n. n+1 -> n transition)
$$ \le \max \Big(g_P(S_n) + g_C(S_n), \frac{\overline{\mathbb{E}} V_{n+1}^P + \overline{\mathbb{E}} V_{n+1}^C}{1 + r}\Big) $$
$$ \le \max \Big(g_P(S_n), \frac{\overline{\mathbb{E}} V_{n+1}^P }{1 + r}\Big) + \max \Big(g_C(S_n), \frac{\overline{\mathbb{E}} V_{n+1}^C}{1 + r}\Big) = V_n^P + V_n^C$$

# Exercise 4.2


$\Delta_1(H) = \frac{V_2(HH) - V_2(HT)}{S_2(HH) - S_2(HT)} = -\frac{1}{12}$

$\Delta_1(T) = \frac{V_2(TH) - V_2(TT)}{S_2(TH) - S_2(TT)} = -\frac{1}{12}$

$\Delta_0 = \frac{V_1(H) - V_1(T)}{S_1(H) - S_1(T)} = -0.43$

$\tau(HH) = \infty, \tau(HT) = 2, \tau(TH)=\tau(TT) = 1$


* first borrow money and buy 0.43 shares 

* If first coin toss is T then borrow exercise put at time 1

* If H then borrow money and buy 1/12 shares

* Then depending on coin toss agent should exercise his put or let it expire. He will pay his debt because he hedged on the previous step (1/12 shares)

# Exercise 4.3

In [9]:
S = {\
    "": 4, \
    "H": 8, \
    "T": 2, \
    "HH": 16,\
    "HT": 4, \
    "TH": 4, \
    "TT": 1, \
    "HHH": 32,\
    "HHT": 8, \
    "HTH": 8, \
    "THH": 8, \
    "HTT": 2, \
    "THT":2, \
    "TTH":2, \
    "TTT": 0.5\
}

In [22]:
def g(x):
    s = 0.0
    for i in range(len(x)+ 1):
        s += S[x[:i]]
    return 4 - s / (len(x) + 1) 

In [31]:
for i in S.keys():
    print("G({})".format(i), g(i))

G() 0.0
G(H) -2.0
G(T) 1.0
G(HH) -5.333333333333334
G(HT) -1.333333333333333
G(TH) 0.6666666666666665
G(TT) 1.6666666666666665
G(HHH) -11.0
G(HHT) -5.0
G(HTH) -2.0
G(THH) -0.5
G(HTT) -0.5
G(THT) 1.0
G(TTH) 1.75
G(TTT) 2.125


$V(THT) = 1.0$

$V(TTH) = 1.75$

$V(TTT) = 2.125$

None zero V

$V_1(TT) = 5/3$

$V_1(TH) = 2/3$

$V_1(T) = 1$

$$V_0 = 0.4$$

So if first coin toss results in H then stopping time is $=\infty$, else it's $=1$ 